---
title: "Exploratory analysis"
author: "Ghada Nouairia & William Wu"
date: "02/07/2024"
output:
  bookdown::html_document2:
    toc: yes
    toc_depth: 2
    toc_float:
      collapse: yes
    fig_caption: yes
    number_sections: no
    code_folding: "hide"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r Read the libraries, message=FALSE, warning=FALSE}
library(tidyverse)
library(rstatix)
library(factoextra)
library(ChAMP)
library(RColorBrewer)
library(limma)
library(ComplexHeatmap)
```

```{r Load the data, message=FALSE, warning=FALSE}
setwd("~/projects/PSC_multiomics")
metabolites <- read_csv("results/metabolite_preprocessed.csv")
proteins <- read_csv("results/protein_preprocessed.csv")
miRNA <- read_csv("results/miRNA_preprocessed.csv") 
metadata <- read_csv("results/metadata_preprocessed.csv")
protein_metadata <- read_csv("data/metadata_OlinkID.csv") %>%
  pivot_longer(cols = c(-1), names_to = "proteins") %>%
  pivot_wider(names_from = c(1))
```

```{r shapiro-wilk_function, message=FALSE, warning=FALSE}
apply_shapiro <- function(data, variables, significance) {
# 1. Running the Shapiro Wilk test
results <- data %>%
  shapiro_test(variables) 
# 2. Filtering for the significant p-values 
results %>% 
  filter(p < significance) %>% 
  nrow()
}
```

```{r t-test_function, message=FALSE, warning=FALSE}
apply_ttest <- function(data, grp1, grp2) {
# 1. Calculating p-values
data %>% 
  select(-1) %>% 
  apply(2, function(x) {
    samp1 <- x[grp1]
    samp2 <- x[grp2]
    test_result <- t.test(samp1, samp2)
    return(test_result$p.value)
  })
}
```

```{r mwu_fold_change_function, message=FALSE, warning=FALSE}
apply_mwu <- function(data, grp1, grp2) {
# 1. Calculating p-values
mwu_p <- data %>% 
  select(-1) %>% 
  apply(2, function(x) {
    samp1 <- x[grp1]
    samp2 <- x[grp2]
    test_result <- wilcox.test(samp1, samp2)
    return(test_result$p.value)
  })

# 2. Calculating fold change
fold_change <- data %>% 
  select(-1) %>% 
  apply(2, function(x){
    samp1 <- x[grp1]
    samp2 <- x[grp2]
    # test_result <- mean(log2(x[seq_along(grp1)])) - mean(log2(x[(length(grp1) + 1):ncol(data)]))
    test_result <- mean(samp1) - mean(samp2)
    return(test_result)
  })

# !!! Previously incorrect mean(log2(psc) - mean(log2(control)))

# 3. Combining to table
variables <- data %>% 
  select(-1) %>% 
  names()

mwu_table <- tibble(
  variable = variables,
  p.value = mwu_p,
  adjusted_p.value = p.adjust(mwu_p, method = "BH"),
  fold_change = fold_change
)}
```

```{r volcano_function, message=FALSE, warning=FALSE}
plot_volcano <- function(data, significance) {
  data %>% 
    ggplot(
      aes(
        x = fold_change,
        y = -log10(p.value),
        color = p.value < significance
      )) +
      geom_point() +
    labs(
      x = "Fold change (log2)",
      y = "P-value (-log10)",
      color = str_glue("p < {significance}")
    )
}
```

```{r density_function, message=FALSE, warning=FALSE}
plot_densities <- function(data, title) {
data %>% 
  select(2:ncol(data)) %>%
  pivot_longer(
    everything(),
    names_to = "variable",
    values_to = "value"
  ) %>% 
  ggplot(mapping = aes(
    x = value,
    y = after_stat(scaled),
    color = variable
  )) +
    geom_density() +
    guides(color = "none") +
    labs(
      x = "Value",
      y = "Density",
      title = title
  )}
```

# Proteins
## Data description
The protein data is analysed for 36 samples, 33 patients and 3 healthy donors.

Method:
Olink’s proximity extension assay (PEA) is used in which each protein are bound by two antibodies with tail DNA oligonucleotides. Upon binding the oligonucleotides hybridize and are extended by DNA polymerase forming a barcode that can be read by qPCR or NGS.
The PEA was done with Olink Target 96, 7 panels of proteins with 92 proteins each, total of 644 proteins. The panels are “CELL REGULATION”, “DEVELOPMENT”, “IMMUNE RESPONSE”, “INFLAMMATION”, “METABOLISM”, “ONCOLOGY II”, “ONCOLOGY III”, “ORGAN DAMAGE”.

Preprocessing:
Only one sample did not pass the quality control, sample ID 30 for “ONCOLOGY III”. Several proteins did not pass quality control and were removed (failed in 8% or more samples, see proteins_preprocessing.R).
The data comes in the unit of normalized protein expression (NPX) values on a log2 scale. They are normalized to interplate controls. 

Links
Olink Target 96: https://olink.com/products/olink-target-96 Olink preprocessing for qPCR: https://7074596.fs1.hubspotusercontent-na1.net/hubfs/7074596/05-white%20paper%20for%20website/1096-olink-data-normalization-white-paper.pdf Panel proteins: https://www.bioxpedia.com/olink-proteomics/#:~:text=The%20readout%20for%20most%20Olink,equals%20a%20high%20protein%20concentration.

## Data analysis {.tabset}
### Initial exploration
```{r proteins_density_plot}
# Plotting the samples/individuals
proteins %>%
  select(-patient_id) %>%
  t() %>%
  limma::plotDensities()
# 
# Plotting the features/variables
# proteins %>%
#   select(-patient_id) %>%
#   limma::plotDensities()
```

Checking sample and data for outliers
```{r proteins_box_plot, message=FALSE, warning=FALSE}
# Plotting the samples/individuals
proteins %>% 
  select(-patient_id) %>% 
  t() %>% 
  boxplot(
    col = palette()[-1],
    main = "Protein distribtution", ylab = "Protein concentration ")

# Plotting the features/variables
proteins %>% 
  select(-patient_id) %>% 
  boxplot(
    col = palette()[-1],
    main = "Protein distribtution", ylab = "Protein concentration ")
```
### MSD
```{r proteins_mds_plot_1, message=FALSE, warning=FALSE}
# Plotting the features/variables
proteins %>%
  select(-patient_id) %>%
  limma::plotMDS()
```

### Shapiro-Wilks test
The number of proteins that do not satisfy normality at p = 0.05 is: 
```{r proteins_shapiro-wilk_test, message=FALSE, warning=FALSE}
# Applying Shapirot_Wilks test for normality
variables <- proteins %>% 
  select(-patient_id) %>% 
  names()
apply_shapiro(proteins, variables, 0.05)
```

### Mann-Whitney U test and Volcano plot
```{r proteins_mwu_test, message=FALSE, warning=FALSE}
mwu <- apply_mwu(proteins, grp1 = c(1:33), grp2 = c(34:36))
```

The number of significantly differentially expressed proteins at p < 0.05 is `r mwu %>% filter(p.value <= 0.05) %>% nrow()`, of which `r mwu %>% filter(p.value <= 0.05 & fold_change > 0) %>% nrow()` are upregulated and `r mwu %>% filter(p.value <= 0.05 & fold_change < 0) %>% nrow()` are downregulated.
```{r table, message=FALSE, warning=FALSE}
datatable(mwu %>%
            inner_join(protein_metadata %>%
                         dplyr::select(proteins, Assay),
                       by = c("variable" = "proteins")) %>%
                         dplyr::select(-variable) %>%
                         relocate(Assay),
          filter =  "top",
          extensions = 'Buttons', 
    options = list(dom = 'Bfrtip', 
                   buttons = c('excel', "csv")))
```

```{r proteins_volcano_plot, message=FALSE, warning=FALSE}
# pdf(file = "results/volcano_prot_controls.pdf")
plot_volcano(mwu, 0.05) 
 # while (!is.null(dev.list()))  dev.off()
```

### SVD 
We check the plausible confounders and the studied effects in the data
```{r proteins_svd, message=FALSE, warning=FALSE, results='hide'}
beta1 <- proteins %>% 
  # dplyr::slice(1:33) %>%       ### no Control
  dplyr::select(-patient_id) %>%
  t() %>% 
  as.data.frame() 
# beta1 <- cbind(beta1[1:11], beta1[13], beta1[29:32])  ####no NoIBD and no CCA

pd1 <- metadata %>% 
  # dplyr::slice(1:33) %>%
  # filter(ibd_binary == "1" & cca_binary == "0") %>%
  select(cca_binary, crohn_or_uc, ibd_binary, alp, alp_binary, bilirubin, bilirubin_binary, fibrosis_binary, sex, age, age_cat, bmi, bmi_cat, group, auto_anb, alat, overlap_aih, steroids_other_medic, jaudice, pruritis, fever_chill, fatigue, no_symptoms) %>%
  as.data.frame()

# pdf(file = "results/illustrations/svd_proteins.pdf")
champ.SVD(beta = beta1, pd = pd1)
# while (!is.null(dev.list()))  dev.off()
```

### MSD multiple dimensions {.tabset}
#### Effect of age category
Using MDS plots to check the grouping of the data according to the confouders found by SVD: colors represent age categories: No much effect of age categories
```{r proteins_mds_plot, message=FALSE, warning=FALSE}
pal <- brewer.pal(8,"Dark2")

proteins %>%
  dplyr::select(-patient_id) %>%
  # dplyr::slice(1:33) %>%
  t() %>%
  plotMDS(
    top=1000,
    gene.selection="common",
    dim.plot = c(1, 2),
    col=pal[factor(metadata$age_cat[1:33])])

proteins %>%
  dplyr::select(-patient_id) %>%
  # dplyr::slice(1:33) %>%
  t() %>%
  plotMDS(
    top=1000,
    gene.selection="common",
    dim.plot = c(1, 3),
    col=pal[factor(metadata$age_cat[1:33])])

proteins %>%
  dplyr::select(-patient_id) %>%
  # dplyr::slice(1:33) %>%
  t() %>%
  plotMDS(
    top=1000,
    gene.selection="common",
    dim.plot = c(2, 3),
    col=pal[factor(metadata$age_cat[1:33])])
```

#### Effect of groups
Looking at the groups (early PSC, CCA, IBD, Advanced and progressing): there is some grouping but not good enough: some feature selection may be needed.
```{r group_effect3, warning=FALSE}
proteins %>%
  dplyr::select(-patient_id) %>%
  dplyr::slice(1:33) %>%
  t() %>%
  plotMDS(
    top=1000,
    gene.selection="common",
    dim.plot = c(1, 2),
    col=pal[factor(metadata$group[1:33])])

proteins %>%
  dplyr::select(-patient_id) %>%
  dplyr::slice(1:33) %>%
  t() %>%
  plotMDS(
    top=1000,
    gene.selection="common",
    dim.plot = c(1, 3),
    col=pal[factor(metadata$group[1:33])])

proteins %>%
  dplyr::select(-patient_id) %>%
  dplyr::slice(1:33) %>%
  t() %>%
  plotMDS(
    top=1000,
    gene.selection="common",
    dim.plot = c(2, 3),
    col=pal[factor(metadata$group[1:33])])
```


### {-}

## {-}

# miRNA
## Data description
The miRNA data is analysed for 36 samples, 33 patients and 3 healthy donors.

Method:
NGS

Preprocessing:
The readings are mapped to known human miRNA from the miRBase data base. !(The data output comes out in the form of counts - no modification). The facility also provides mature miRNA log transformed values (that we use here). Constant values of this dataset were removed.

## Data analysis {.tabset}
### Initial explorataion
Checking sample's outliers
```{r miRNA_box_plot, message=FALSE, warning=FALSE}
# Plotting the samples/individuals
miRNA %>% 
  select(-patient_id) %>% 
  t() %>% 
  boxplot(
    col = palette()[-1],
    main = "miRNA distribtution", ylab = "miRNA concentration ")
```

### Shapiro-Wilks test
miRNA expression data often does not follow a normal distribution, even after log transformation. The number of miRNA that do not satisfy normality at p = 0.05 is: 1420
```{r miRNA_shapiro-wilk_test, message=FALSE, warning=FALSE}
# Applying Shapiro Wilks test for normality
variables <- miRNA %>% 
  select(-patient_id) %>% 
  names()
apply_shapiro(miRNA, variables, 0.05)
```

### Mann-Whitney U test and Volcano plot
```{r miRNA_mwu_test, message=FALSE, warning=FALSE}
mwu <- apply_mwu(miRNA, grp1 = c(1:33), grp2 = c(34:36))
    
datatable(mwu,
          filter =  "top",
          extensions = 'Buttons', 
    options = list(dom = 'Bfrtip', 
                   buttons = c('excel', "csv")))
```

The number of significantly differentially expressed miRNA at p < 0.05 is `r mwu %>% filter(p.value <= 0.05) %>% nrow()`, of which `r mwu %>% filter(p.value <= 0.05 & fold_change > 0) %>% nrow()` are upregulated and `r mwu %>% filter(p.value <= 0.05 & fold_change < 0) %>% nrow()` are downregulated.
```{r miRNA_volcano_plot, message=FALSE, warning=FALSE}
# pdf(file = "results/volcano_miRNA_controls.pdf")
plot_volcano(mwu, 0.05) + scale_color_manual(values = alpha(c("black", "red"), 0.5)) 
# while (!is.null(dev.list()))  dev.off()
```

### SVD 
We check the plausible confounders and the studied effects in the data
```{r miRNA_svd, message=FALSE, warning=FALSE, results='hide'}
beta1 <- miRNA %>% 
  dplyr::slice(1:33) %>%
  dplyr::select(-patient_id) %>%
  t() %>% 
  as.data.frame()

pd1 <- metadata %>% 
  dplyr::slice(1:33) %>%
  select(cca_binary, crohn_or_uc, ibd_binary, alp, alp_binary, bilirubin, bilirubin_binary, fibrosis_binary, sex, age, age_cat, bmi, bmi_cat, group, auto_anb, alat, overlap_aih, steroids_other_medic, jaudice, pruritis, fever_chill, fatigue, no_symptoms) %>%
  as.data.frame()
# pdf(file = "results/svd_miRNA.pdf")
champ.SVD(beta = beta1, pd = pd1)
 # while (!is.null(dev.list()))  dev.off()
```

### MDS plot of groups
Using MDS plots to check the grouping of the data 
```{r miRNA_mds_plot, message=FALSE, warning=FALSE}
pal <- brewer.pal(8,"Dark2")

miRNA %>%
  dplyr::select(-patient_id) %>%
  dplyr::slice(1:33) %>%
  t() %>%
  plotMDS(
    top=1000,
    gene.selection="common",
    dim.plot = c(1, 2),
    col=pal[factor(metadata$group[1:33])])

miRNA %>%
  dplyr::select(-patient_id) %>%
  dplyr::slice(1:33) %>%
  t() %>%
  plotMDS(
    top=1000,
    gene.selection="common",
    dim.plot = c(1, 3),
    col=pal[factor(metadata$group[1:33])])

miRNA %>%
  dplyr::select(-patient_id) %>%
  dplyr::slice(1:33) %>%
  t() %>%
  plotMDS(
    top=1000,
    gene.selection="common",
    dim.plot = c(2, 3),
    col=pal[factor(metadata$group[1:33])])
```

## {-}

# Metabolites
## Data description
The metabolite data is analysed for 45 samples, 33 patients and 12 healthy donors.

Method:
Ultra performance liquid chromatography + Tandem mass spectroscopy.

Preprocessing:
The values are log transformed (base unknown). The missing values are imputated with the minimum observed value of that respective compound. The data is normalized so that the median/medians are 1.0000 and all other values are proportional.

## Data analysis {.tabset}
```{r log, message=FALSE, warning=FALSE}
log2_metabolites <- cbind(
  metabolites[1], 
  log2(metabolites[2:length(metabolites)]))
```

### Initil plots
```{r metabolites_density_plot, message=FALSE, warning=FALSE}
# Plotting the samples/individuals
log2_metabolites %>%
  select(-patient_id) %>%
  t() %>%
  limma::plotDensities()
# 
# # Plotting the features/variables
# proteins %>% 
#   select(-patient_id) %>% 
#   limma::plotDensities()
```

Checking sample's outliers
```{r metabolites_box_plot, message=FALSE, warning=FALSE}
# Plotting the samples/individuals
log2_metabolites %>% 
  select(-patient_id) %>% 
  t() %>% 
  boxplot(
    col = palette()[-1],
    main = "Protein distribtution", ylab = "Protein concentration ")

# Plotting the features/variables
log2_metabolites %>% 
  select(-patient_id) %>% 
  boxplot(
    col = palette()[-1],
    main = "Protein distribtution", ylab = "Protein concentration ")
```

### Shapiro-Wilks test
The number of metabolites that do not satisfy normality at p = 0.05 is: 524
```{r metabolites_shapiro-wilk_test, message=FALSE, warning=FALSE}
# Applying Shapirot_Wilks test for normality
variables <- log2_metabolites %>% 
  select(-patient_id) %>% 
  names()
apply_shapiro(log2_metabolites, variables, 0.05)
```

### Mann-Whitney U test and Volcano plot
```{r metabolites_mwu_test, message=FALSE, warning=FALSE}
mwu <- apply_mwu(log2_metabolites, grp1 = c(1:33), grp2 = c(34:45))
datatable(mwu,
          filter =  "top",
          extensions = 'Buttons', 
    options = list(dom = 'Bfrtip', 
                   buttons = c('excel', "csv")))
```
The number of significantly differentially expressed miRNA at p < 0.05 is `r mwu %>% filter(p.value <= 0.05) %>% nrow()`, of which `r mwu %>% filter(p.value <= 0.05 & fold_change > 0) %>% nrow()` are upregulated and `r mwu %>% filter(p.value <= 0.05 & fold_change < 0) %>% nrow()` are downregulated.
```{r metabolites_volcano_plot, message=FALSE, warning=FALSE}
# pdf(file = "results/volcano_metabolomics.pdf")
plot_volcano(mwu, 0.05) + scale_color_manual(values = alpha(c("black", "red"), 0.5)) 
 # while (!is.null(dev.list()))  dev.off()
```

### SVD
We check the plausible confounders and the studied effects in the data
```{r metabolites_svd, message=FALSE, warning=FALSE, results='hide'}
beta1 <- log2_metabolites %>% 
  dplyr::slice(1:36) %>% 
  dplyr::select(-patient_id) %>%
  t() %>% 
  as.data.frame()

pd1 <- metadata %>% 
  # dplyr::slice(1:33) %>%
  select(cca_binary, crohn_or_uc, ibd_binary, alp, alp_binary, bilirubin, bilirubin_binary, fibrosis_binary, sex, age, age_cat, bmi, bmi_cat, group, auto_anb, alat, overlap_aih, steroids_other_medic, jaudice, pruritis, fever_chill, fatigue, no_symptoms) %>%
  as.data.frame()

# pdf(file = "results/svd_metabolomics.pdf")
champ.SVD(beta = beta1, pd = pd1)
 # while (!is.null(dev.list()))  dev.off()
```

### MDS multiple dimensions {.tabset}
Using MDS plots to check the grouping of the data according to the confouders found by SVD. Age effect is clearly visible in the data. It may be necessary to adjust for it or to do some feature selection.

#### Age category effect
```{r metabolites_mds_plot, message=FALSE, warning=FALSE}
pal <- brewer.pal(8,"Dark2")

log2_metabolites %>%
  dplyr::select(-patient_id) %>%
  # dplyr::slice(1:33) %>%
  t() %>%
  plotMDS(
    top=1000,
    gene.selection="common",
    dim.plot = c(1, 2),
    col=pal[factor(metadata$age_cat[1:33])])

log2_metabolites %>%
  dplyr::select(-patient_id) %>%
  # dplyr::slice(1:33) %>%
  t() %>%
  plotMDS(
    top=1000,
    gene.selection="common",
    dim.plot = c(1, 3),
    col=pal[factor(metadata$age_cat[1:33])])

log2_metabolites %>%
  dplyr::select(-patient_id) %>%
  # dplyr::slice(1:33) %>%
  t() %>%
  plotMDS(
    top=1000,
    gene.selection="common",
    dim.plot = c(2, 3),
    col=pal[factor(metadata$age_cat[1:33])])
```
Looking at the groups (early PSC, CCA, IBD, Advanced and progressing): there is some grouping but not good enough: some feature selection may be needed.

#### Effect of groups
```{r group_effect2, warning=FALSE, message=FALSE}
log2_metabolites %>%
  dplyr::select(-patient_id) %>%
  dplyr::slice(1:33) %>%
  t() %>%
  plotMDS(
    top=1000,
    gene.selection="common",
    dim.plot = c(1, 2),
    col=pal[factor(metadata$group[1:33])])

log2_metabolites %>%
  dplyr::select(-patient_id) %>%
  dplyr::slice(1:33) %>%
  t() %>%
  plotMDS(
    top=1000,
    gene.selection="common",
    dim.plot = c(1, 3),
    col=pal[factor(metadata$group[1:33])])

log2_metabolites %>%
  dplyr::select(-patient_id) %>%
  dplyr::slice(1:33) %>%
  t() %>%
  plotMDS(
    top=1000,
    gene.selection="common",
    dim.plot = c(2, 3),
    col=pal[factor(metadata$group[1:33])])
```

### {-}

## {-}

## Unsupervised clustering of the data {.tabset}
### Metabolomics
Scaled, Best k = 4
```{r kmeans_metab, warning=FALSE, message=FALSE}
res_km_r <- eclust(scale(log2_metabolites %>% dplyr::select(-patient_id)), "kmeans")
fviz_gap_stat(res_km_r$gap_stat)
# pdf(file = "results/Illustrations/metabolomics_kmean_K4.pdf")
res_km_r <- eclust(scale(log2_metabolites %>% dplyr::select(-patient_id)), "kmeans", k = 4) 
# while (!is.null(dev.list()))  dev.off()

res_km_r <- eclust(scale(log2_metabolites), FUNcluster = "hclust", k = 4, graph = TRUE)
fviz_cluster(res_km_r, scale(log2_metabolites %>% dplyr::select(-patient_id)), ellipse.type = "ellipse", palette = c("#0b5313", "#ec4dd8", "blue", "orange"), ggtheme = theme_minimal())
```

Non scaled
```{r kmean_metab, warning=FALSE, message=FALSE}
res_km_r <- eclust(log2_metabolites %>% dplyr::select(-patient_id), "kmeans")
fviz_gap_stat(res_km_r$gap_stat)
# pdf(file = "results/Paper_illustratios/kmean_lasso_grpcolor.pdf")
res_km_r <- eclust(log2_metabolites%>% dplyr::select(-patient_id), "kmeans", k = 4) 
# while (!is.null(dev.list()))  dev.off()

res_km_r <- eclust(log2_metabolites%>% dplyr::select(-patient_id), FUNcluster = "hclust", k = 4, graph = TRUE)
fviz_cluster(res_km_r, log2_metabolites%>% dplyr::select(-patient_id), ellipse.type = "ellipse", palette = c("#0b5313", "#ec4dd8", "blue", "orange"), ggtheme = theme_minimal())
```

### Proteomics data
```{r kmeans_prot, warning=FALSE, message=FALSE}
res_km_r <- eclust(scale(proteins %>% dplyr::select(-patient_id)), "kmeans")
fviz_gap_stat(res_km_r$gap_stat)
# pdf(file = "results/Illustrations/proteomics_kmean_k4.pdf")
res_km_r <- eclust(scale(proteins %>% dplyr::select(-patient_id)), "kmeans", k = 4) 
# while (!is.null(dev.list()))  dev.off()

res_km_r <- eclust(scale(proteins %>% dplyr::select(-patient_id)), FUNcluster = "hclust", k = 4, graph = TRUE)
fviz_cluster(res_km_r, scale(proteins %>% dplyr::select(-patient_id)), ellipse.type = "ellipse", palette = c("#0b5313", "#ec4dd8", "blue", "orange"), ggtheme = theme_minimal())
```

### miRNA data
```{r kmeans_miRNA, warning=FALSE, message=FALSE}
res_km_r <- eclust(scale(miRNA %>% dplyr::select(-patient_id)), "kmeans")
fviz_gap_stat(res_km_r$gap_stat)
# pdf(file = "results/Illustrations/transcriptomics_kmean_k4.pdf")
res_km_r <- eclust(scale(miRNA %>% dplyr::select(-patient_id)), "kmeans", k = 3)
# while (!is.null(dev.list()))  dev.off()

res_km_r <- eclust(scale(miRNA %>% dplyr::select(-patient_id)), FUNcluster = "hclust", k = 4, graph = TRUE)
fviz_cluster(res_km_r, scale(miRNA %>% dplyr::select(-patient_id)), ellipse.type = "ellipse", palette = c("#0b5313", "#ec4dd8", "blue", "orange"), ggtheme = theme_minimal())
```

## {-}

```{r zoom-in_heatmap, message=FALSE, fig.cap="Protein expression is differential between groups"}
# pdf(file = "results/Paper_illustratios/heatmap1_EN_univ_sel_ptn.pdf")
# Heatmap(as.matrix(miRNA %>% dplyr::select(-patient_id)), 
#         row_labels = rownames(miRNA %>% dplyr::select(-patient_id)),
#         column_labels = colnames(miRNA %>% dplyr::select(-patient_id)),
#         cluster_rows = TRUE,
#         row_names_max_width = unit(6, "cm"),
#         cluster_columns = FALSE,
#         row_names_rot = 15,
#         row_names_gp = gpar(fontsize = 3)) # Text size for row names
# while (!is.null(dev.list()))  dev.off()
```


