---
title: "WGCNA analysis of PSC patients"
author: "William Wu"
date: "2024-08-07"
output: 
  html_document:
    toc: TRUE
    toc_float: TRUE
    code_folding: "hide"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Background

We use the Weighted Gene Correlation Network Analysis (WGCNA) to analyse how the variables of our omics-data can be grouped together. The scaled (Z-score) and concatenated omics-data is analysed most thouroughly while each omics is analysed separately in a simpler way. Some chunks are commented out but left in the script to make it easier for someone who has done the Horvath tutorial to understand and modify this code. The graphical settings using the base-package graphics are unfortunately not optimized.

In summary the network model divides the variables into groups called modules based on the correlation between the variables. Thereafter, the module eigenvalues are related to clinical traits and the correlation values and p-values are calculated. For the most interesting module-trait pairs the gene significance (variable-trait correlation) and module membership (variable-module eigengene correlation) are calculated. Lastly all variables of the network model along with their gene significance and module membership are exported for further enrichment analysis.

**WGCNA Application to Proteomic and Metabolomic Data Analysis**

Originally WGCNA was developed only for gene expression. However, with proper missing data imputation and normalization, proteins and metabolites can be studied as well.

The correlation between variables is either positive or negative. In the signed analysis the negative correlation is not considered while the unsigned analysis calculates the absolute value of the correlation making positive and negative correlation equal. This article recommends to construct a signed network.

Article: <https://www.sciencedirect.com/science/article/pii/S0076687916302890>

**Horvath tutorial**

Down below is a short overview of the each step in the Horvath tutorial that can be accessed in the following link below. The most important functions are written down along with their most important arguments. For this analysis the steps 0, 1, 2.b, 3, and 5 are used in a modified way. The most interesting modules are exported.

Tutorial: <https://www.dropbox.com/scl/fo/4vqfiysan6rlurfo2pbnk/h?rlkey=thqg8wlpdn4spu3ihjuc1kmlu&e=1&dl=0>

Parameters:\
0. Filtering the omics-data for variables with low variance\
nearZeroVar() - freqCut =, uniqueCut =\
Comment: The function nearZeroVar() is used identically as in our mixOmics analysis

1.  Removing samples\
    cutreeStatic() - cutHeight = - Removes outliers\
    Comment: This step is solely used for initial visualization and not for removing outliers due to our low number of samples

2.a. Creating an automatic network\
blockwiseModules() - power = - Set soft threshold power with pickSoftThreshold()\
blockwiseModules() - TOMType = - Choose network type (unsigned or signed)\
blockwiseModules() - minModuleSize = - Choose minimal size of modules\
blockwiseModules() - deepSplit = - Choose sensitivity for cluster splitting (default = 2)\
Comment: This step was initially used but later abandoned in favor of the manual network

2.b. Creating a manual network\
pickSoftThreshold()\
adjacency() - Correlation matrix between all variables to the power of the soft threshold\
TOMsimilarity() - Transformation of the adjacency matrix to reduce noise and spurious associations\
hclust() - Create gene tree\
cutreeDynamic() - Create modules\
moduleEigengenes() - Calculate module eigengenes to create dendrogram\
mergeCloseModules() - Merge modules with similar expression profiles according to dendrogram\
Comment: This step is not optimized, i. e. the hyperparameters have not been iterated or motivated thouroughly

3.  Relating the modules with clinical traits\
    labeledHeatmap() - Choose which clinical traits to investigate for which modules (gene significance = variable-trait correlation; module membership = variable-module eigengene correlation)\
    Comment: This step is the most important one and is extended so that the most interesting module-trait pairs are exported into csv.files for further analysis

4.  Annotating with gene ontology\
    Subset interesting modules for annotation and enrichment analysis\
    Comment: This step is done in our enrichment_analysis script

5.  Visualizing the network model\
    Visualize gene network with heatmap and dendrogram\
    Visualize module network for a clinical variable with heatmap and dendrogram\
    Comment: This step is mainly important for visualising the entire network model

6.  Exporting the network model\
    Export gene network for visualization with external tools\
    Comment: This step is skipped since we are not using the external tools

## Read the libraries

## Read the libraries

```{r read_the_libraries, message=FALSE, warning=FALSE}
# Load essential libraries for data manipulation, WGCNA analysis, modeling, and parallel processing
library(tidyverse)     # For data manipulation and visualization
library(WGCNA)         # For weighted gene co-expression network analysis
library(caret)         # For preprocessing and feature selection
library(igraph)        # For network visualization and graph operations
library(doParallel)    # For enabling parallel computatio
library(multiGSEA)
library(dplyr)
```

```{r}
library(conflicted)    # Handles conflicts between functions from different packages

# Prefer using specific functions from selected packages in case of conflict
conflict_prefer("slice", "dplyr")
conflict_prefer("rename", "dplyr")
conflict_prefer("cor", "WGCNA")
conflict_prefer("select","dplyr")
conflict_prefer("reduce","purrr")
```

```{r prallel}
# Initialize a parallel backend using 4 cores
cl <- makeCluster(4)  # Adjust number of cores depending on your system
registerDoParallel(cl)
```

## Set prequisites

```{r prequisites, message=FALSE, warning=FALSE}
# Ensure strings are not automatically converted to factors
options(stringsAsFactors = FALSE)

# Enable multithreading in WGCNA to speed up computations
enableWGCNAThreads()
```

## Load the data

The omics data comes in a uniform format with each row representing a patient/sample and each column representing a compound/feature/variable except the first column which contains the patient ID's. The metadata has been cleaned for the variables of interest so that they are represented by new binary columns. See the preprocessing script for details.

```{r load_data, message=FALSE, warning=FALSE}
# Load preprocessed datasets for each omic layer and the associated metadata
metabolites <- read_csv("../data/metabolite_preprocessed.csv")
proteins <- read_csv("../data/protein_preprocessed.csv")
miRNA <- read_csv("../data/miRNA_preprocessed.csv")
metadata <- read_csv("../data/metadata_preprocessed.csv")


medata_OlinkID <- read_csv("../data/metadata_OlinkID.csv")
medata_metabolites <- read_csv("medata_metabolites_kegg_filled.csv")
```

## Prepare the data

The patient ID column is set as the rownames to keep track of the rows.\
The metabolite data was probably not log2 transformed but only scaled with 1 as the median and all other values proportional.\
-\> The metabolite data needs to be log2 transformed The numerical columns of interest are selected in the metadata.

```{r format_data, message=FALSE, warning=FALSE}
# Keep first 33 patients and remove miRNAs with low variance
miRNA <- miRNA %>% 
  slice(1:33)
miRNA <- miRNA[,-nearZeroVar(miRNA, uniqueCut = 30)]  # Removes 1920 low-variance miRNAs

# Keep first 33 patients for proteins
proteins <- proteins %>% 
  slice(1:33)

# Log2-transform and filter metabolite data
metabolites <- metabolites %>% 
  log2() %>% 
  slice(1:33)
metabolites <- metabolites[,-nearZeroVar(metabolites, uniqueCut = 30)]  # Removes 29 low-variance metabolites

# Select only binary traits of interest from metadata
metadata <- metadata %>% 
  dplyr::select(cca_binary, ibd_binary, fibrosis_binary, alp_binary, bilirubin_binary) %>% 
  slice(1:33)




# Rename the binary traits for readability
formatted_names <- c("CCA", "IBD", "Fibrosis", "ALP", "Bilirubin")
```

We scale (Z-score) and concatenate the omics data.

```{r concatenate_data}
# Combine scaled miRNA, protein, and metabolite data into a single dataframe
multi_omics <- cbind(
  scale(miRNA[2:ncol(miRNA)]),        # Remove ID column and scale miRNA data
  scale(proteins[2:ncol(proteins)]),  # Remove ID column and scale protein data
  scale(metabolites[2:ncol(metabolites)]) # Remove ID column and scale metabolite data
) %>% 
  as.data.frame() %>% 
  select_if(~ n_distinct(.x) > 1)      # Remove columns with no variability

```

This function checks if there are any columns with too many missing values. Result: There are no such columns.

```{r check_missing_values}
# Check for missing values or genes/samples with too many NAs
# WGCNA recommends using this function before proceeding with analysis

gsg = goodSamplesGenes(multi_omics, verbose = 3)  # Check multi-omics data
gsg$allOK                                          # Should return TRUE if all data is suitable

gsg = goodSamplesGenes(miRNA[2:ncol(miRNA)], verbose = 3)  # Check miRNA data separately
gsg$allOK

gsg = goodSamplesGenes(proteins[2:ncol(proteins)], verbose = 3)  # Check protein data separately
gsg$allOK

gsg = goodSamplesGenes(metabolites[2:ncol(metabolites)], verbose = 3)  # Check metabolite data separately
gsg$allOK

```

# Multi-omics {.tabset}

## 1. Data input

This initial plot groups the samples with hierarchical clustering to check for outliers.

```{r multi_omics_visualise_samples}

# Perform hierarchical clustering on the combined multi-omics dataset
sampleTree = hclust(dist(multi_omics), method = "average")  # Use average linkage for clustering

# Create a dendrogram plot of the sample tree and save as a PDF
pdf(file = "../results/WGCNA/hclust.pdf", width = 12, height = 9)  # Adjust dimensions as needed
par(cex = 0.6)                    # Set character expansion for labels
par(mar = c(2, 8, 2, 0))          # Set plot margins

# Plot the dendrogram to visualize sample clustering
plot(sampleTree, 
     main = "Sample clustering to detect outliers", 
     sub = "", xlab = "", 
     cex.lab = 1.5, cex.axis = 1.5, cex.main = 2)

# Optionally add a red horizontal line to help identify outliers
abline(h = 75, col = "red")
```

```{r multi_omics_filter_data}
##**Skipped**: This function divides the data into the clusters that can be seen in the plot above. Then only the main cluster ##is kept and the data is referred to as fmulti_omics. Any variables that are constant in the filtered dataset are removed.
# # Determine cluster under the line
# clust = cutreeStatic(sampleTree, cutHeight = 75, minSize = 10)
# table(clust)
# # clust 1 contains the samples we want to keep.
# keepSamples = (clust==1)
# fmulti_omics = multi_omics[keepSamples, ]
# 
# # Remove the constant variables in our filtered dataset
# fmulti_omics <- fmulti_omics %>% 
#     select_if(~ n_distinct(.x) > 1)
# 
# nGenes = ncol(fmulti_omics)
# nSamples = nrow(fmulti_omics)
```

```{r multi_omics_filter_metadata}
##**Skipped**: The metadata need to be matched to the main cluster that was kept and it is referred to as fmetadata.
# # Form a data frame analogous to expression data that will hold the clinical traits.
# Samples = rownames(fmulti_omics);
# traitRows = match(Samples, metadata$patient_id);
# fmetadata = metadata[traitRows, ];
# fmetadata <- fmetadata %>% column_to_rownames(var = "patient_id")
# collectGarbage();
```

The heatmap below show the hierarchical clustering of the samples with corresponding metadata where white means low, red means high, and grey means missing entry.

```{r multi_omics_visualise_traits}
# Convert binary clinical traits to color-coded matrix for visualization
# White = low (0), Red = high (1), Grey = missing
traitColors = numbers2colors(metadata, signed = FALSE)

# Save dendrogram + trait heatmap to PDF
pdf(file = "../results/WGCNA/sampleClustering_correlation_groups.pdf", width = 12, height = 9)
plotDendroAndColors(sampleTree, traitColors,
                    groupLabels = formatted_names, 
                    main = "Sample dendrogram and trait heatmap")
```

Intrestingly the Early and NoIBD groups are clustered together.

## 2.a Automatic network construction

We set an interval of soft thresholding powers and plot the scale independence and mean connectivity as a function of those powers. The scale independence reaches the plateau around power = 4.

```{r multi_omics_automatic_choose_power}
# Re-initialize parallel backend (ensure it's running)
cl <- makeCluster(4)  # Adapt number of cores to your machine
registerDoParallel(cl)

# Define a range of candidate powers for soft-thresholding
powers = c(c(1:10), seq(from = 12, to = 20, by = 2))

# Analyze scale-free topology fit and connectivity at each power
sft = pickSoftThreshold(multi_omics, powerVector = powers, verbose = 5)

# Save the soft-threshold analysis plots to PDF
pdf(file = "../results/WGCNA/Automatic_soft_threshold_analysis.pdf", width = 12, height = 9)
par(mfrow = c(1, 2))  # Plot two graphs side by side
cex1 = 0.9

# Plot scale-free topology fit index
plot(sft$fitIndices[,1], 
     -sign(sft$fitIndices[,3]) * sft$fitIndices[,2],
     xlab = "Soft Threshold (power)", 
     ylab = "Scale Free Topology Model Fit, signed R²", 
     type = "n", main = "Scale independence")
text(sft$fitIndices[,1], 
     -sign(sft$fitIndices[,3]) * sft$fitIndices[,2],
     labels = powers, cex = cex1, col = "red")
# Optional threshold line for R² (e.g. 0.90)
# abline(h = 0.90, col = "red")

# Plot mean connectivity for each power
plot(sft$fitIndices[,1], sft$fitIndices[,5],
     xlab = "Soft Threshold (power)", 
     ylab = "Mean Connectivity", 
     type = "n", main = "Mean connectivity")
text(sft$fitIndices[,1], sft$fitIndices[,5], 
     labels = powers, cex = cex1, col = "red")
```

We create a network with our variables divided into 16 modules with outliers in module 0.

```{r multi_omics_automatic_create_network}
# Construct network and detect modules (unsupervised WGCNA)
net = blockwiseModules(multi_omics, power = 4,
                       TOMType = "unsigned", 
                       minModuleSize = 30,               # Minimum size of each module
                       reassignThreshold = 0, 
                       mergeCutHeight = 0.25,            # Merge threshold for similar modules
                       numericLabels = TRUE, 
                       pamRespectsDendro = FALSE, 
                       saveTOMs = TRUE,                  # Save TOM matrices for future use
                       saveTOMFileBase = "multi_omicsTOM",
                       verbose = 3)

# Show the distribution of module sizes (by label)
table(net$colors)
```

We create the dendrogram for our network.

```{r multi_omics__automatic_network_dendrogram}
# Set graphic window dimensions
sizeGrWindow(12, 9)

# Convert module numeric labels to color names
mergedColors = labels2colors(net$colors)

# Plot the gene dendrogram and color-coded module assignments
pdf(file = "../results/WGCNA/automatic_cluster_dendogram.pdf", width = 12, height = 9)
plotDendroAndColors(net$dendrograms[[1]], 
                    mergedColors[net$blockGenes[[1]]],
                    "Module colors",
                    dendroLabels = FALSE, hang = 0.03,
                    addGuide = TRUE, guideHang = 0.05)
```

We save some of the results from our network for further analysis. These include the module distribution, its colors, the module eigenvalues, and the dendrogram.

```{r multi_omics_automatic_save_network}
# Extract module assignments, color codes, and eigengenes
moduleLabels = net$colors                       # Numeric module labels
moduleColors = labels2colors(net$colors)        # Corresponding color labels
MEs = net$MEs                                    # Module eigengenes (summary expression profiles)
geneTree = net$dendrograms[[1]]                 # Dendrogram of all features  
```

## 2.b Manual network construction

We set an interval of soft thresholding powers and plot the scale independence and mean connectivity as a function of those powers. The scale independence reaches the plateau around power = 7.

```{r multi_omics_manual_choose_power}
# Choose a refined set of candidate soft-thresholding powers
powers = c(c(1:5), seq(from = 5, to = 10, by = 1))

# Run network topology analysis to identify optimal power
sft = pickSoftThreshold(multi_omics, powerVector = powers, verbose = 5)

# Plot scale independence and connectivity
pdf(file = "../results/WGCNA/Manual_soft_threshold_analysis.pdf", width = 12, height = 9)
par(mfrow = c(1,2))
cex1 = 0.9

# Plot scale-free topology model fit
plot(sft$fitIndices[,1], -sign(sft$fitIndices[,3]) * sft$fitIndices[,2],
     xlab = "Soft Threshold (power)", ylab = "Scale Free Topology Fit, signed R²", 
     type = "n", main = "Scale independence")
text(sft$fitIndices[,1], -sign(sft$fitIndices[,3]) * sft$fitIndices[,2], 
     labels = powers, cex = cex1, col = "red")

# Plot mean connectivity
plot(sft$fitIndices[,1], sft$fitIndices[,5],
     xlab = "Soft Threshold (power)", ylab = "Mean Connectivity", 
     type = "n", main = "Mean connectivity")
text(sft$fitIndices[,1], sft$fitIndices[,5], 
     labels = powers, cex = cex1, col = "red")

# Manually set the soft-threshold power for adjacency calculation
softPower = 7
```

We calculate the correlation between each variable in a pair-wise manner.

```{r multi_omics_manal_adjacency}
# Compute the adjacency matrix with the chosen power
adjacency = adjacency(multi_omics, power = softPower)

# Compute topological overlap matrix (TOM) and dissimilarity
TOM = TOMsimilarity(adjacency)
dissTOM = 1 - TOM

```

We visualize the dendrogram.

```{r multi_omics_manual_variables_dendrogram}
# Hierarchical clustering on TOM-based dissimilarity
geneTree = hclust(as.dist(dissTOM), method = "average")

# Plot the gene dendrogram
pdf(file = "../results/WGCNA/manual_soft_threshold_analysis.pdf", width = 12, height = 9)
plot(geneTree, xlab = "", sub = "", main = "Variable clustering on TOM-based dissimilarity",
     labels = FALSE, hang = 0.04)
```

We create 11 modules with the 0th module representing outliers.

```{r multi_omics_manual_modules}
# Identify modules using dynamic tree cut
minModuleSize = 30
dynamicMods = cutreeDynamic(dendro = geneTree, distM = dissTOM,
                            deepSplit = 2, pamRespectsDendro = FALSE,
                            minClusterSize = minModuleSize)
table(dynamicMods)
```

We visualize the dendrogram again, this time with the corresponding module colors below.

```{r multi_omics_manual_modules_dendrogram}
# Assign module colors
dynamicColors = labels2colors(dynamicMods)
table(dynamicColors)

# Visualize modules on the dendrogram
pdf(file = "../results/WGCNA/manual_cluster_dendogram_man_14.pdf", width = 12, height = 9)
plotDendroAndColors(geneTree, dynamicColors, "Dynamic Tree Cut",
                    dendroLabels = FALSE, hang = 0.03,
                    addGuide = TRUE, guideHang = 0.05,
                    main = "Variable dendrogram and module colors")
```

We visualize the clustering of the module eigengenes.

```{r multi_omics_manual_eigengenes_network}
# Calculate module eigengenes and their dissimilarity
MEList = moduleEigengenes(multi_omics, colors = dynamicColors)
MEs = MEList$eigengenes
MEDiss = 1 - cor(MEs)
METree = hclust(as.dist(MEDiss), method = "average")

# Plot clustering of module eigengenes
pdf(file = "../results/WGCNA/module_merging.pdf", width = 12, height = 9)
plot(METree, main = "Clustering of module eigengenes", xlab = "", sub = "")
abline(h = 0.15, col = "red")  # Optional: show merge threshold

MEDissThres = 0.15
merge = mergeCloseModules(multi_omics, dynamicColors, cutHeight = MEDissThres, verbose = 3)
mergedColors = merge$colors
mergedMEs = merge$newMEs
```

We are not merging any modules and therefore set the cutting line (MEDissThres) below the lowerst branches in the chunk above.

We visualize the cut tree with original and merged modules - No changes since we did not merge any modules.

```{r multi_omics_manual_network_dendrogram}
# Plot comparison of original and merged modules
pdf(file = "../results/WGCNA/modules_merged_dendogram.pdf", width = 12, height = 9)
plotDendroAndColors(geneTree, cbind(dynamicColors, mergedColors),
                    c("Dynamic Tree Cut", "Merged dynamic"),
                    dendroLabels = FALSE, hang = 0.03,
                    addGuide = TRUE, guideHang = 0.05)
```

We skip this step since the variables have already been set.

```{r save_modules}
### Rename to moduleColors
moduleColors = mergedColors
### Construct numerical labels corresponding to the colors
colorOrder = c("grey", standardColors(50));
moduleLabels = match(moduleColors, colorOrder)-1;
MEs = mergedMEs;
### Save module colors and labels for use in subsequent parts
save(MEs, moduleLabels, moduleColors, geneTree, file = "../results/WGCNA/modules_info.csv")
```

```{r}
# Correlate each variable with each module eigengene
MMandPvalue = corAndPvalue(multi_omics, MEs, use = "p")
MM = MMandPvalue$cor
pval = MMandPvalue$p

# Save module membership with p-values
moduleMembershipDF = data.frame(
  Variable    = colnames(multi_omics),
  ModuleColor = moduleColors,
  MM,
  pval
)
write.csv(moduleMembershipDF, file = "../results/WGCNA/module_membership_pvalues.csv", row.names = FALSE)
```

## 3. Relating modules to traits

We calculate the ME values for the modules and subsequently calculate the correlation and p-value in regards to the metadata.

```{r multi_omics_calculate_module_eigengenes}
moduleColors = mergedColors

# Number of variables (genes/features) and number of samples
nGenes = ncol(multi_omics)
nSamples = nrow(multi_omics)

# Retrieve precomputed module eigengenes
MEs = mergedMEs

# Compute Pearson correlations between module eigengenes and traits
moduleTraitCor = cor(MEs, metadata, use = "p")
moduleTraitPvalue = corPvalueStudent(moduleTraitCor, nSamples)
```

We create a heatmap of the correlation and p-values that were calculated above. Here we can see which modules are significant for which trait.

```{r multi_omics_plot_modules_traits}
# Format correlation and p-value matrix as a labeled text matrix
textMatrix = paste(signif(moduleTraitCor, 2), "\n(",
                   signif(moduleTraitPvalue, 1), ")", sep = "")
dim(textMatrix) = dim(moduleTraitCor)

# Generate and export the heatmap as PDF
pdf(file = "../results/WGCNA/manual_module_trait_pearson_heatmap.pdf", wi = 9, he = 6)
par(mar = c(5, 10, 4, 4))  # bottom, left, top, right
labeledHeatmap(Matrix = moduleTraitCor,
               xLabels = formatted_names,
               yLabels = names(MEs),
               ySymbols = names(MEs),
               colorLabels = FALSE,
               colors = blueWhiteRed(50),
               textMatrix = textMatrix,
               setStdMargins = FALSE,
               cex.text = 0.5,
               zlim = c(-1,1),
               main = "Module-trait relationships (Pearson)")
```

**Manual network interesting module-trait pairs** From this figure I identify the following pairs to be of interest (p \< 0,001):\
fibrosis: brown, green\
fibrosis_binary: brown, green\
alp: brown\
alp_binary: brown, green\
bilirubin: purple, red, brown\
bilirubin_binary: purple, brown, green

In the following sections we calculate the module mebership and gene significance for the inteeresting module-trait pairs and then export his information into csv files.

## 4. Focus on each trait

## Scatterplot of module membership (MM) vs gene significance (GS)

```{r}
scatter_MM_GS <- function(module, traitCol, topN = 10, color) {
  
  # --- Quick checks ---
  stopifnot(traitCol %in% colnames(metadata))
  stopifnot(module %in% moduleColors)

  # --- Format trait as dataframe ---
  traitDF <- data.frame(metadata[[traitCol]])
  names(traitDF) <- traitCol

  # --- Compute MM and GS correlations ---
  modNames <- substring(names(MEs), 3)
  geneMM   <- as.data.frame(cor(multi_omics, MEs, use = "p"))
  geneGS   <- as.data.frame(cor(multi_omics, traitDF, use = "p"))

  # --- Subset for variables in selected module ---
  colIdx <- match(module, modNames)
  moduleGenes <- moduleColors == module

  MMabs <- abs(geneMM[moduleGenes, colIdx])
  GSabs <- abs(geneGS[moduleGenes, 1])
  varNames <- names(multi_omics)[moduleGenes]

  # --- Rank top variables by MM × GS score ---
  score <- MMabs * GSabs
  topIdx <- order(score, decreasing = TRUE)[1:min(topN, length(score))]

  # --- Plot scatterplot ---
  verboseScatterplot(
    x = MMabs, y = GSabs,
    xlab = paste("Module Membership in", module, "module"),
    ylab = paste("Gene significance for", traitCol),
    main = "Module membership vs. gene significance",
    col = color,
    pch = 16, cex = 0.6,
    cex.main = 1.2, cex.lab = 1.2, cex.axis = 1.2
  )

  # Optional: annotate topN points (currently commented out)
  # text(MMabs[topIdx], GSabs[topIdx] + 0.02,
  #      labels = varNames[topIdx], pos = 3, cex = 0.5, col = "black")
}
```

```{r}
## Function to compute and visualize module activity scores by phenotype group
multiomics_score <- function(module_color = "yellow",
                             phenotype = "CCA",
                             module_df = moduleMembershipDF,
                             omics_df = multi_omics,
                             meta_df = metadata,
                             do_plot = TRUE) {

  # --- 1. Select variables in target module ---
  vars <- module_df %>%
    dplyr::filter(ModuleColor == module_color) %>%
    dplyr::pull(Variable)

  if (length(vars) == 0)
    stop("No variables found for module ‘", module_color, "’.")

  omics_filt <- omics_df %>% dplyr::select(dplyr::all_of(vars))

  # --- 2. Create binary trait indicator (case vs control) ---
  pheno_bin_col <- paste0(tolower(phenotype), "_binary")
  if (!pheno_bin_col %in% names(meta_df))
    stop("Column ‘", pheno_bin_col, "’ is missing in metadata.")

  is_case <- meta_df[[pheno_bin_col]] == 1
  omics_case <- omics_filt[is_case, , drop = FALSE]
  omics_control <- omics_filt[!is_case, , drop = FALSE]

  # --- 3. Combine with group labels ---
  df_all <- dplyr::bind_rows(
    as.data.frame(omics_case)    %>% dplyr::mutate(group = phenotype),
    as.data.frame(omics_control) %>% dplyr::mutate(group = paste0("No_", phenotype))
  )

  # --- 4. Z-score normalization (per variable) ---
  df_scaled <- df_all %>%
    dplyr::mutate(dplyr::across(where(is.numeric), ~ scale(.)[, 1]))

  # --- 5. Compute average module score per sample ---
  df_scores <- df_scaled %>%
    dplyr::rowwise() %>%
    dplyr::mutate(score = mean(dplyr::c_across(where(is.numeric)), na.rm = TRUE)) %>%
    dplyr::ungroup() %>%
    dplyr::select(group, score)

  # --- 6. Sample counts and label positioning ---
  df_n <- df_scores %>%
    dplyr::group_by(group) %>%
    dplyr::summarise(n = dplyr::n(),
                     y_pos = max(score, na.rm = TRUE) + 0.05,
                     .groups = "drop")

  # --- 7. Statistical tests ---
  wilcox_res <- stats::wilcox.test(score ~ group, data = df_scores, exact = FALSE)
  cliff_res  <- effsize::cliff.delta(score ~ group, data = df_scores)
  pval <- formatC(wilcox_res$p.value, digits = 3, format = "g")

  # --- 8. Optional boxplot visualization ---
  if (isTRUE(do_plot)) {
    suppressMessages({
      p <- ggplot2::ggplot(df_scores, ggplot2::aes(group, score, fill = group)) +
        ggplot2::stat_boxplot(geom = "errorbar", width = .25) +
        ggplot2::geom_boxplot(width = 0.6, outlier.shape = NA) +
        ggplot2::geom_jitter(width = .1, size = 2, alpha = .6) +
        ggplot2::geom_text(data = df_n,
                           ggplot2::aes(x = group, y = y_pos,
                                        label = paste0("n = ", n)),
                           vjust = 0, fontface = "bold", size = 4) +
        ggplot2::expand_limits(y = max(df_n$y_pos) + 0.1) +
        ggplot2::theme_bw() +
        ggplot2::theme(panel.grid = ggplot2::element_blank(),
                       aspect.ratio = 2,
                       legend.position = "none",
                       plot.title = ggplot2::element_text(hjust = 0.5)) +
        ggplot2::labs(x = "", y = "Average z-score (module activity)",
                      title = paste0(phenotype, " - Wilcoxon test (", module_color, "), p = ", pval))
      print(p)
    })
  }

  # --- 9. Return score and test results ---
  list(scores = df_scores,
       wilcox = wilcox_res,
       cliff  = cliff_res)
}
```

### Fibrosis

```{r}
# Visualize MM vs GS for brown and green modules (Fibrosis)
pdf(file = "../results/WGCNA/fibrosis_module_membership.pdf", wi = 9, he = 6)
scatter_MM_GS("brown", "fibrosis_binary", color = "brown")
scatter_MM_GS("green", "fibrosis_binary", color = "forestgreen")
```

```{r}
# Compare module scores between fibrosis and non-fibrosis groups
pdf(file = "../results/WGCNA/fibrosis_modules_boxplot.pdf", wi = 9, he = 6)
multiomics_score("brown", "Fibrosis", moduleMembershipDF, multi_omics, metadata)
multiomics_score("green", "Fibrosis", moduleMembershipDF, multi_omics, metadata)
```

### ALP

```{r}
pdf(file = "../results/WGCNA/alp_module_membership.pdf", wi = 9, he = 6)
scatter_MM_GS("brown", "alp_binary", color = "brown")
scatter_MM_GS("green", "alp_binary", color = "forestgreen")
```

```{r}
pdf(file = "../results/WGCNA/alp_modules_boxplot.pdf", wi = 9, he = 6)
multiomics_score("brown", "ALP", moduleMembershipDF, multi_omics, metadata)
multiomics_score("green", "ALP", moduleMembershipDF, multi_omics, metadata)
```

### Bilirubin

```{r}
pdf(file = "../results/WGCNA/bilirubin_module_membership.pdf", wi = 9, he = 6)
scatter_MM_GS("brown", "bilirubin_binary", color = "brown")
scatter_MM_GS("green", "bilirubin_binary", color = "forestgreen")
```

```{r}
pdf(file = "../results/WGCNA/bilirubin_modules_boxplot.pdf", wi = 9, he = 6)
multiomics_score("brown", "Bilirubin", moduleMembershipDF, multi_omics, metadata)
multiomics_score("green", "Bilirubin", moduleMembershipDF, multi_omics, metadata)
```

### CCA

```{r}
pdf(file = "../results/WGCNA/CCA_module_membership.pdf", wi = 9, he = 6)
scatter_MM_GS("yellow", "cca_binary", color = "gold")
```

```{r}
pdf(file = "../results/WGCNA/CCA_modules_boxplot.pdf", wi = 9, he = 6)
multiomics_score("yellow", "CCA", moduleMembershipDF, multi_omics, metadata)
```

### IBD

```{r}
pdf(file = "../results/WGCNA/IBD_module_membership.pdf", wi = 9, he = 6)
scatter_MM_GS("yellow", "ibd_binary", color = "gold")
scatter_MM_GS("brown", "ibd_binary", color = "brown")

```

```{r}
pdf(file = "../results/WGCNA/IBD_modules_boxplot.pdf", wi = 9, he = 6)
multiomics_score("yellow", "IBD", moduleMembershipDF, multi_omics, metadata)
multiomics_score("brown", "IBD", moduleMembershipDF, multi_omics, metadata)
```

## 5. Global visualisation

We create a heatmap that shows the correlation between all of our variables. The bright spots are our modules.

```{r multi_omics_plot_variable_dendrogram_heatmap}
# Recompute the TOM matrix
dissTOM = 1 - TOMsimilarityFromExpr(multi_omics, power = 7)
plotTOM = dissTOM^7
diag(plotTOM) = NA  # Remove diagonal for clarity

pdf(file = "../results/WGCNA/WGCNA_tomplot.pdf", wi = 9, he = 6)
TOMplot(plotTOM, geneTree, moduleColors, main = "Network heatmap and dendrogram")
```

```{r}
# Prepare trait names and compute correlations with modules
traitNames <- c("cca_binary", "ibd_binary", "alp_binary", "fibrosis_binary", "bilirubin_binary")
stopifnot(all(rownames(mergedMEs) == rownames(metadata)))
traits <- metadata[, traitNames, drop = FALSE]
MET <- orderMEs(cbind(mergedMEs, traits))

# Compute distance matrix (1 - |cor|) and cluster
dissME <- 1 - abs(cor(MET, use = "p"))
tree <- hclust(as.dist(dissME), method = "average")

# Rename trait labels for readability
labels_raw <- colnames(MET)
labels_custom <- sapply(labels_raw, function(name) {
  if (grepl("^ME", name)) return(name)
  name <- gsub("_binary", "", name)
  switch(name,
         alp = "ALP",
         cca = "CCA",
         ibd = "IBD",
         fibrosis = "Fibrosis",
         bilirubin = "Bilirubin",
         name)
})
tree$labels <- labels_custom

# Plot dendrogram of eigengenes and traits
pdf(file = "../results/WGCNA/WGCNA_tomplot.pdf", wi = 9, he = 6)
plot(tree,
     main = "Dendrogram of module eigengenes and traits\n(distance = 1 - |cor|)",
     xlab = "", sub = "", hang = 0.03, cex = 1)
dev.off()
```

```{r}
# Compute signed correlation matrix and rename dimensions
corMAT <- cor(MET, use = "p")
prettyNames <- c("CCA", "IBD", "ALP", "Fibrosis", "Bilirubin")
names(prettyNames) <- traitNames

renameFun <- function(x) ifelse(x %in% traitNames, prettyNames[x], x)
rownames(corMAT) <- renameFun(rownames(corMAT))
colnames(corMAT) <- renameFun(colnames(corMAT))

# Order rows/columns by hierarchical clustering
tree <- hclust(as.dist(1 - abs(corMAT)), method = "average")
ord <- tree$order
corMAT <- corMAT[ord, ord]

# Format text for cell labels
textMatrix <- signif(corMAT, 2)
dim(textMatrix) <- dim(corMAT)


pdf(file = "../results/WGCNA/Heatmap_moduleXtrait.pdf", wi = 9, he = 6)
par(mar = c(8, 8, 4, 4))  # bottom, left, top, right

labeledHeatmap(
  Matrix = corMAT,
  xLabels = colnames(corMAT),
  yLabels = rownames(corMAT),
  ySymbols = rownames(corMAT),
  colorLabels = FALSE,
  colors = blueWhiteRed(50),
  textMatrix = textMatrix,
  cex.text = 0.55,
  cex.lab = 1,
  xLabelsAngle = 90,
  setStdMargins = FALSE,
  zlim = c(-1, 1),
  main = "Signed correlation heatmap – modules & traits"
)
dev.off()
```


## 6. Mapping



```{r}
library(multiMiR)

### Column selection ----------------------------------------------------------
gene_cols <- grep("^hsa[-_]",  colnames(multi_omics), value = TRUE,
                  ignore.case = TRUE)       # miRNA features
prot_cols <- grep("^OID",      colnames(multi_omics), value = TRUE)  # proteins
met_cols  <- grep("^(X)?[0-9]+$", colnames(multi_omics), value = TRUE) # metabolites

# Olink OID → UniProt mapping
oid_map <- medata_OlinkID %>%
  filter(OlinkID == "Uniprot ID") %>%
  pivot_longer(-OlinkID, names_to = "oid", values_to = "uniprot") %>%
  filter(uniprot != "")

# Metabolite COMP ID → HMDB / KEGG mapping
met_map <- medata_metabolites %>%
  select(`COMP ID`, HMDB, KEGG) %>%
  mutate(cid_char = as.character(`COMP ID`),
         id_final = if_else(HMDB != "" & !is.na(HMDB), HMDB, KEGG)) %>%
  filter(id_final != "") %>%
  select(cid_char, id_final)

# ---------------------------------------------------------------------------
# 2. Query multiMiR (validated targets only) ---------------------------------
# ---------------------------------------------------------------------------
qry <- get_multimir(
  mirna   = gene_cols,
  org     = "hsa",
  table   = "validated",   # miRecords, miRTarBase, TarBase
  summary = FALSE          # keep each DB separate
)

mirna2gene <- qry@data %>%
  filter(database == "mirtarbase") %>%
  select(mature_mirna_id, target_symbol) %>%
  distinct()
```



##7. Module enrichment + DIABLO for KEGGEnrichissement par module + diablo pour KEGG

### Combination of both outputs

```{r}
library(readr)
library(dplyr)
library(ggVennDiagram)
library(ggplot2)

library(readr)
library(dplyr)
library(ggVennDiagram)
library(ggplot2)

module_disease_analysis <- function(color, disease) {

  # Step 1: load mixOmics selection for the disease
  disease_file <- paste0("../results/mixOmics_", disease, "_binary.csv")
  disease_data <- read_csv(disease_file, show_col_types = FALSE)
  compounds_list <- disease_data$compounds

  # Step 2: variables belonging to the selected module
  filtered_module <- moduleMembershipDF %>%
    filter(ModuleColor == color)
  variables_list <- rownames(filtered_module)

  # Step 3: Venn diagram -----------------------------------------------------
  venn_plot <- ggVennDiagram(
    list(Compounds = compounds_list,
         Variables = variables_list),
    label = "count"
  ) +
    theme_void()

  print(venn_plot)

  # Step 4: merge both lists, keep only present variables
  merged_list <- union(compounds_list, variables_list)
  filtered_multiomics <- multi_omics[, intersect(colnames(multi_omics),
                                                 merged_list)]

  return(filtered_multiomics)
}

# Example call
resultat <- module_disease_analysis("yellow", "cca")
```

### Split variables by omics layer

```{r}
### Column selection (after filtering) ----------------------------------------
gene_cols <- grep("^hsa[-_]",  colnames(resultat), value = TRUE, ignore.case = TRUE)
prot_cols <- grep("^OID",      colnames(resultat), value = TRUE)
met_cols  <- grep("^(X)?[0-9]+$", colnames(resultat), value = TRUE)

print(length(gene_cols))
print(length(prot_cols))
print(length(met_cols))
```

### Enrichissement

```{r}
library(clusterProfiler)
library(dplyr); library(stringr); library(tibble)
library(multiMiR); library(graphite); library(multiGSEA)
library(dplyr); library(stringr); library(purrr)
library(ggplot2); library(scico)

library(readr);  library(dplyr); library(purrr);  library(tidyr)
library(clusterProfiler); library(enrichplot);    library(ggplot2)
library(metap);  library(org.Hs.eg.db);           library(ggVennDiagram)
```


```{r}
library(org.Hs.eg.db)

################ Metabolites ##################################################
kegg_vec <- medata_metabolites %>%
  filter(as.character(`COMP ID`) %in% met_cols) %>%
  filter(!is.na(KEGG), !KEGG %in% c("NA", "undefined")) %>%
  pull(KEGG) %>%
  unique()

kegg_split  <- strsplit(kegg_vec, ",")   # split on commas
kegg_flat   <- unlist(kegg_split)
kegg_clean  <- trimws(kegg_flat)
kegg_unique <- unique(kegg_clean)

ora_kegg_cpd <- enrichKEGG(
  gene          = kegg_unique,   # C‑numbers (e.g. "C00022")
  organism      = "cpd",
  keyType       = "kegg",
  pvalueCutoff  = 0.05,
  pAdjustMethod = "BH",
  maxGSSize     = 500
)

################ Proteins #####################################################
proteine_uniprot <- oid_map %>%
  filter(oid %in% prot_cols)

prot_interest <- proteine_uniprot$uniprot

ora_kegg_prot <- enrichKEGG(
  gene           = prot_interest,  # UniProt accessions
  organism       = "hsa",
  keyType        = "uniprot",
  pvalueCutoff   = 0.05,
  pAdjustMethod  = "BH"
)

################ miRNA ########################################################
mirna_symbol <- mirna2gene %>%
  filter(mature_mirna_id %in% gene_cols)

gene_interest <- mirna_symbol$target_symbol |>
  na.omit() |>
  unique()

eg <- bitr(
  gene_interest,
  fromType = "SYMBOL",
  toType   = "ENTREZID",
  OrgDb    = org.Hs.eg.db
)

ora_kegg_gene <- enrichKEGG(
  gene            = eg$ENTREZID,
  keyType         = "kegg",
  organism        = "hsa",
  pAdjustMethod   = "BH",
  pvalueCutoff    = 0.05,
  maxGSSize       = 500
)

```


### Merge p‑values

```{r}
library(dplyr)
library(purrr)

###############################################################################
# 0) Libraries already loaded above                                           #
###############################################################################

normalize_pathway <- function(id) sub("^(hsa|map)", "", id)

prot_df <- ora_kegg_prot@result %>%
  as_tibble() %>%
  transmute(path_num   = normalize_pathway(ID),
            description = Description,
            prot_padj   = p.adjust)

gene_df <- ora_kegg_gene@result %>%
  as_tibble() %>%
  transmute(path_num   = normalize_pathway(ID),
            description = Description,
            gene_padj   = p.adjust)

met_df <- ora_kegg_cpd@result %>%
  as_tibble() %>%
  transmute(path_num   = normalize_pathway(ID),
            description = Description,
            met_padj    = p.adjust)

###############################################################################
# Combine p‑values (Stouffer method)                                          #
###############################################################################
combinePvalues <- function(df, col_pattern = "_padj") {
  cols <- grep(col_pattern, colnames(df))
  apply(df[, cols], 1, function(row) {
    row <- row[!is.na(row)]
    row[row == 0] <- .Machine$double.xmin
    if (length(row) >= 2) return(sumz(row)$p)   # Stouffer
    if (length(row) == 1) return(row)
    NA_real_
  })
}

merged_kegg <- list(gene_df, prot_df, met_df) %>%
  reduce(full_join, by = "path_num") %>%
  mutate(description = coalesce(description.x, description.y, description)) %>%
  select(path_num, description, ends_with("_padj")) %>%
  mutate(
    p_comb_stouffer     = combinePvalues(cur_data()),
    p_comb_stouffer_adj = p.adjust(p_comb_stouffer, method = "BH")
  )

sig <- merged_kegg %>% filter(p_comb_stouffer_adj <= 0.05)

###############################################################################
# 4) Retrieve real gene / protein / compound sets                             #
###############################################################################
split_slash <- function(x) ifelse(is.na(x) | x == "", character(0),
                                  strsplit(x, "/", fixed = TRUE))

genesets_gene <- ora_kegg_gene@result %>%
  as_tibble() %>%
  transmute(path_num = normalize_pathway(ID),
            gs       = map(geneID, split_slash))

genesets_prot <- ora_kegg_prot@result %>%
  as_tibble() %>%
  transmute(path_num = normalize_pathway(ID),
            gs       = map(geneID, split_slash))

genesets_met <- ora_kegg_cpd@result %>%
  as_tibble() %>%
  transmute(path_num = normalize_pathway(ID),
            gs       = map(geneID, split_slash))

genesets_all <- bind_rows(genesets_gene, genesets_prot, genesets_met) %>%
  group_by(path_num) %>%
  summarise(geneSets = list(unique(unlist(gs))),
            Count    = lengths(geneSets),
            .groups  = "drop")

###############################################################################
# 5) Complete enrichResult table                                              #
###############################################################################
sig2 <- sig %>% left_join(genesets_all, by = "path_num")

```

### Treeplot without omics priority


```{r}
################################################################################
# 4) Convert to enrichResult object                                           #
###############################################################################
res_df <- sig2 %>%
  transmute(
    ID          = path_num,
    Description = description,
    pvalue      = p_comb_stouffer,
    p.adjust    = p_comb_stouffer_adj,
    qvalue      = p_comb_stouffer_adj,
    geneID      = map_chr(geneSets, paste, collapse = "/"),
    Count,
    GeneRatio   = paste0(Count, "/", Count),  # placeholders
    BgRatio     = GeneRatio
  ) %>%
  as.data.frame()

er <- new("enrichResult",
          result        = res_df,
          geneSets      = setNames(sig2$geneSets, sig2$path_num),
          pvalueCutoff  = 1,
          pAdjustMethod = "none",
          qvalueCutoff  = 1,
          organism      = "UNKNOWN",
          keytype       = "UNKNOWN",
          readable      = FALSE)

```

```{r}
# 1) Rebuild geneSets from er@result -----------------------------------------
gs_list <- setNames(
  lapply(er@result$geneID, function(s)
    strsplit(s, "/", fixed = TRUE)[[1]] |> unique()),
  er@result$ID
)

er@geneSets            <- gs_list
er@result$Count        <- lengths(gs_list)
er@result$GeneRatio    <- paste0(er@result$Count, "/", er@result$Count)
er@result$BgRatio      <- er@result$GeneRatio  # placeholder

###############################################################################
# 2) Manual Jaccard similarity matrix                                         #
###############################################################################
ids <- names(gs_list); n <- length(ids)
jac <- matrix(0, n, n, dimnames = list(ids, ids))
for (i in seq_len(n)) {
  gi <- gs_list[[i]]
  for (j in i:n) {
    gj <- gs_list[[j]]
    inter <- length(intersect(gi, gj))
    uni   <- length(unique(c(gi, gj)))
    val   <- ifelse(uni == 0, 0, inter / uni)
    jac[i, j] <- val; jac[j, i] <- val
  }
}
attr(er, "termsim") <- jac

###############################################################################
# 3b) Replace IDs by “Description” labels for display ------------------------
###############################################################################
lab_vec <- with(er@result, setNames(Description, ID))
rownames(jac)        <- colnames(jac) <- lab_vec[rownames(jac)]
names(gs_list)       <- lab_vec[names(gs_list)]
er@result$ID         <- lab_vec[er@result$ID]
attr(er, "termsim")  <- jac          # jac is now relabelled

###############################################################################
# 3) Treeplot -----------------------------------------------------------------
###############################################################################
p <- treeplot(
  er,
  showCategory   = 30,
  cluster.params = list(method = "average"),
  label_format   = 60
)

ggsave("treeplot_kegg_manual.pdf",
       p, width = 16.5, height = 11.7, units = "in")
```


### Treeplot with omics priority


```{r}
###############################################################################
# 0) Term selection & ranking for treeplot ------------------------------------
###############################################################################
sig2_ranked <- sig2 %>%
  mutate(
    nOmics = (!is.na(gene_padj)) + (!is.na(prot_padj)) + (!is.na(met_padj))
  ) %>%
  filter(p_comb_stouffer_adj < 0.05) %>%
  arrange(desc(nOmics), p_comb_stouffer_adj)

top30_ids <- head(sig2_ranked$path_num, 40)


res_df <- sig2 %>%                                # keep *all* terms for `er`
  transmute(
    ID          = path_num,
    Description = description,
    pvalue      = p_comb_stouffer,
    p.adjust    = p_comb_stouffer_adj,
    qvalue      = p_comb_stouffer_adj,
    geneID      = map_chr(geneSets, paste, collapse = "/"),
    Count,
    GeneRatio   = paste0(Count, "/", Count),       # placeholders
    BgRatio     = GeneRatio
  ) %>%
  as.data.frame()

er <- new("enrichResult",
          result        = res_df,
          geneSets      = setNames(sig2$geneSets, sig2$path_num),
          pvalueCutoff  = 1,
          pAdjustMethod = "none",
          qvalueCutoff  = 1,
          organism      = "UNKNOWN",
          keytype       = "UNKNOWN",
          readable      = FALSE)


gs_list <- setNames(
  lapply(er@result$geneID, \(s) strsplit(s, "/", fixed = TRUE)[[1]] |> unique()),
  er@result$ID
)

er@geneSets         <- gs_list
er@result$Count     <- lengths(gs_list)
er@result$GeneRatio <- paste0(er@result$Count, "/", er@result$Count)
er@result$BgRatio   <- er@result$GeneRatio     # placeholder

###############################################################################
# 2) Manual Jaccard similarity matrix ----------------------------------------
###############################################################################
ids <- names(gs_list); n <- length(ids)
jac <- matrix(0, n, n, dimnames = list(ids, ids))
for (i in seq_len(n)) {
  gi <- gs_list[[i]]
  for (j in i:n) {
    gj <- gs_list[[j]]
    inter <- length(intersect(gi, gj))
    uni   <- length(unique(c(gi, gj)))
    val   <- ifelse(uni == 0, 0, inter / uni)
    jac[i, j] <- val; jac[j, i] <- val
  }
}
attr(er, "termsim") <- jac

###############################################################################
# 3b) Replace IDs by description labels (readability) -------------------------
###############################################################################
lab_vec <- with(er@result, setNames(Description, ID))
rownames(jac)        <- colnames(jac) <- lab_vec[rownames(jac)]
names(gs_list)       <- lab_vec[names(gs_list)]
er@result$ID         <- lab_vec[er@result$ID]
attr(er, "termsim")  <- jac            # similarity matrix relabelled

###############################################################################
# 3) Priority treeplot (top ≈30 terms) ----------------------------------------
###############################################################################
show_vec <- lab_vec[top30_ids] |> unname()

cluster_cols <- c(
  "#2ca02c",  # cluster 1
  "#33caff",  # cluster 2
  "#9467bd",  # cluster 3
  "#d62727",  # cluster 4
  "#ffdf33"   # cluster 5
)

p <- treeplot(
  er,
  showCategory   = show_vec,
  cluster.params = list(
    method = "average",
    n      = 5,                 # number of clusters
    color  = cluster_cols
  ),
  label_format   = 60,
  color          = "p.adjust"   # continuous scale for node colour
)

ggsave("treeplot_kegg_manuals.pdf",
       p, width = 16.5, height = 11.7, units = "in")
```


### Extract omics present in ≥ 75 % of terms per cluster 

```{r}
###############################################################################
# 0) Preparation --------------------------------------------------------------
###############################################################################
library(dplyr)

min_prop <- 0.75          # at least 75 % of terms in the cluster

  

###############################################################################
# 1) Global table: TERM → raw identifiers (genes, proteins, compounds…) -------
###############################################################################
term_to_ids <- setNames(
  lapply(er@result$geneID, \(s) unique(strsplit(s, "/", fixed = TRUE)[[1]])),
  er@result$ID
)


###############################################################################
# 2 + 3) Loop through clusters to get representative IDs ----------------------
###############################################################################
representative_ids_by_cluster <- list()
unique_ids_by_term_by_cluster <- list()

cluster_ids <- p[["data"]] |>
  filter(isTip, !is.na(group), grepl("^cluster_", group)) |>
  pull(group) |>
  unique()

for (cl in cluster_ids) {

  # a) Terms inside the current cluster
  terms <- p[["data"]] |>
    filter(isTip, group == cl, !is.na(label)) |>
    pull(label) |>
    unique()
  if (!length(terms)) next  # empty cluster -> skip

  # b) TERM → id list for this cluster
  ids_by_term <- term_to_ids[terms]

  # c) Representative ids (frequency ≥ min_prop)
  freq <- table(unlist(ids_by_term))
  representative_ids_by_cluster[[cl]] <-
    sort(names(freq)[freq / length(ids_by_term) >= min_prop])

  # d) IDs unique to each term (across all clusters)
  uniq <- lapply(names(ids_by_term), function(term) {
    setdiff(ids_by_term[[term]],
            unlist(term_to_ids[names(term_to_ids) != term]))
  })
  names(uniq) <- names(ids_by_term)
  unique_ids_by_term_by_cluster[[cl]] <- uniq
}



###############################################################################
# 5) Convert numeric gene IDs → miRNA targets; keep other IDs unchanged -------
###############################################################################
geneid_to_symbol <- setNames(eg$SYMBOL, as.character(eg$ENTREZID))
symbol_to_mirnas <- split(mirna_symbol$mature_mirna_id,
                          mirna_symbol$target_symbol)

ids2miRNAs_keep_others <- function(ids) {
  ids           <- as.character(ids)
  id_is_numeric <- grepl("^\\d+$", ids)
  gene_ids      <- ids[id_is_numeric]
  other_ids     <- ids[!id_is_numeric]

  # GeneID → SYMBOL → miRNA
  syms <- geneid_to_symbol[gene_ids] |> na.omit()
  miRs <- unique(unlist(symbol_to_mirnas[syms]))

  unique(c(other_ids, miRs))
}

representative_mixed_by_cluster <-
  lapply(representative_ids_by_cluster, ids2miRNAs_keep_others)
unique_mixed_by_term_by_cluster <-
  lapply(unique_ids_by_term_by_cluster,
         function(clist) lapply(clist, ids2miRNAs_keep_others))

```

```{r}
###############################################################################
# 0) Long-format dataframe for ggplot -----------------------------------------
###############################################################################
library(tidyverse)

df <- imap_dfr(representative_mixed_by_cluster,
               ~tibble(id = .x, cluster = .y))

# Y‑axis ordering: alphabetical then reversed (top‑down)
df$id      <- factor(df$id, levels = rev(sort(unique(df$id))))
df$cluster <- factor(df$cluster, levels = names(representative_mixed_by_cluster))

###############################################################################
# 1) Cluster palette (reuse colours above) ------------------------------------
###############################################################################
cluster_cols <- c(
  cluster_1 = "#2ca02c",
  cluster_2 = "#33caff",
  cluster_3 = "#9467bd",
  cluster_4 = "#d62727",
  cluster_5 = "#ffdf33"
)[levels(df$cluster)]


###############################################################################
# 2) Build and save heat‑map ---------------------------------------------------
###############################################################################
p <- ggplot(df, aes(x = cluster, y = id)) +
  geom_tile(aes(fill = cluster), colour = "white", width = .9, height = .9) +
  scale_fill_manual(values = cluster_cols, guide = "none") +
  labs(
    title = "Omics present in ≥75% of terms within each cluster",
    x = "Cluster",
    y = NULL
  ) +
  theme_minimal(base_size = 12) +
  theme(
    axis.text.x = element_text(angle = 45, hjust = 1),
    panel.grid  = element_blank()
  )

pdf("cluster_heatmap.pdf", width = 10, height = 14)      # tall portrait
print(p)
dev.off()
```




## 8.  Reusable KEGG pipeline function

```{r}
###############################################################################
#  MAIN FUNCTION : multi_module_kegg_pipeline                                 #
###############################################################################

## Helper: build a filtered multi‑omics dataset for the selected modules
multi_module_disease_analysis <- function(modules, disease, show_venn = TRUE) {

  # --- 1. Compounds selected by mixOmics for the disease ---------------------
  disease_file   <- paste0("../results/mixOmics_", disease, "_binary.csv")
  compounds_list <- readr::read_csv(disease_file, show_col_types = FALSE)$compounds

  # --- 2. Variables belonging to all requested modules -----------------------
  variables_list_union <- moduleMembershipDF |>
    dplyr::filter(ModuleColor %in% modules) |>
    rownames()

  # --- 3. Single Venn diagram ------------------------------------------------
  if (show_venn) {
    ven <- ggVennDiagram::ggVennDiagram(
             list(Compounds = compounds_list,
                  Variables = variables_list_union),
             label = "count") +
           ggplot2::theme_void()
    print(ven)
  }

  # --- 4. Return the filtered multi‑omics matrix -----------------------------
  merged_list         <- union(compounds_list, variables_list_union)
  filtered_multiomics <- multi_omics[, intersect(colnames(multi_omics), merged_list)]

  return(filtered_multiomics)
}


# ----------------------------------------------------------------------------- 
# Helper: get KEGG compound name for a single identifier ----------------------
get_kegg_name <- function(id) {
  url <- paste0("https://rest.kegg.jp/list/cpd:", id)   # KEGG “list” endpoint
  res <- GET(url)
  stop_for_status(res)
  txt <- content(res, "text", encoding = "UTF-8")
  # format: "cpd:C00186\tL‑Glutamate\n"
  strsplit(txt, "\t")[[1]][2] |> trimws()
}

# Helper: get UniProt recommended name for a single accession -----------------
get_uniprot_name <- function(id) {
  url <- paste0("https://rest.uniprot.org/uniprotkb/", id, ".json")
  res <- GET(url, accept_json())
  stop_for_status(res)
  j <- fromJSON(content(res, "text", encoding = "UTF-8"))
  j$proteinDescription$recommendedName$fullName$value
}


multi_module_kegg_pipeline <- function(modules,
                                       disease,
                                       p_cutoff = 0.05,
                                       out_pdf  = "treeplot_kegg_manual.pdf",
                                       verbose  = TRUE) {

  ## 0) Load libraries --------------------------------------------------------
  suppressPackageStartupMessages({
    library(readr);  library(dplyr);  library(purrr);  library(tidyr)
    library(clusterProfiler); library(enrichplot);  library(ggplot2)
    library(metap);  library(org.Hs.eg.db);         library(ggVennDiagram)
    library(tibble)   # for tibble::tibble
  })

  ## 1) Create the combined multi‑omics dataset -------------------------------
  if (verbose) message("➜ Step 1: building the multi‑omics dataset …")
  multi_omics_mod <- multi_module_disease_analysis(modules, disease,
                                                   show_venn = verbose)

  ## 2) Split variables (genes / proteins / metabolites) ----------------------
  if (verbose) message("➜ Step 2: splitting genes / proteins / metabolites …")
  gene_cols <- grep("^hsa[-_]",      colnames(multi_omics_mod), value = TRUE,
                    ignore.case = TRUE)
  prot_cols <- grep("^OID",          colnames(multi_omics_mod), value = TRUE)
  met_cols  <- grep("^(X)?[0-9]+$",  colnames(multi_omics_mod), value = TRUE)

  ## 3) Run KEGG over‑representation analyses ---------------------------------
  if (verbose) message("➜ Step 3: running KEGG enrichments …")

  # 3a) Metabolites -----------------------------------------------------------
  kegg_unique <- medata_metabolites %>%
    filter(as.character(`COMP ID`) %in% met_cols,
           !is.na(KEGG), !KEGG %in% c("NA", "undefined")) %>%
    pull(KEGG) |> strsplit(",") |> unlist() |> trimws() |> unique()

  ora_kegg_cpd <- enrichKEGG(gene = kegg_unique, organism = "cpd",
                             keyType = "kegg", pvalueCutoff = 1,
                             pAdjustMethod = "BH", maxGSSize = 500)

  # 3b) Proteins --------------------------------------------------------------
  prot_interest <- oid_map %>%
    filter(oid %in% prot_cols) %>% pull(uniprot) %>% unique()

  ora_kegg_prot <- enrichKEGG(gene = prot_interest, organism = "hsa",
                              keyType = "uniprot", pvalueCutoff = 1,
                              pAdjustMethod = "BH")

  # 3c) miRNA → target genes --------------------------------------------------
  gene_interest <- mirna2gene %>%
    filter(mature_mirna_id %in% gene_cols) %>%
    pull(target_symbol) %>% na.omit() %>% unique()

  eg <- bitr(gene_interest, fromType = "SYMBOL", toType = "ENTREZID",
             OrgDb = org.Hs.eg.db)

  ora_kegg_gene <- enrichKEGG(gene = eg$ENTREZID, organism = "hsa",
                              keyType = "kegg", pvalueCutoff = 1,
                              pAdjustMethod = "BH", maxGSSize = 500)

  ## 4) Meta‑analysis of p‑values (Stouffer) ----------------------------------
  if (verbose) message("➜ Step 4: merging & meta‑analysing p‑values …")

  normalize_pathway <- function(id) sub("^(hsa|map)", "", id)
  tidy_enrich <- function(er, colname) er@result %>%
    as_tibble() %>%
    transmute(path_num   = normalize_pathway(ID),
              description = Description,
              !!colname   := p.adjust)

  prot_df <- tidy_enrich(ora_kegg_prot, "prot_padj")
  gene_df <- tidy_enrich(ora_kegg_gene, "gene_padj")
  met_df  <- tidy_enrich(ora_kegg_cpd,  "met_padj")

  merged_kegg <- list(gene_df, prot_df, met_df) %>%
    reduce(full_join, by = "path_num") %>%
    mutate(description = coalesce(description.x, description.y, description)) %>%
    select(path_num, description, ends_with("_padj")) %>%
    mutate(
      p_comb_stouffer = apply(across(ends_with("_padj")), 1, \(row){
        vals <- row[!is.na(row)]
        vals[vals == 0] <- .Machine$double.xmin
        if (length(vals) >= 2) return(sumz(vals)$p)
        if (length(vals) == 1) return(vals)
        NA_real_
      }),
      p_comb_stouffer_adj = p.adjust(p_comb_stouffer, method = "BH")
    )

  sig <- merged_kegg %>% filter(p_comb_stouffer_adj <= p_cutoff)

  ## 5) Treeplot of significant pathways --------------------------------------
  if (verbose) message("➜ Step 5: generating the treeplot …")

  # 5a) Build geneSets list ---------------------------------------------------
  split_slash <- \(x) ifelse(is.na(x) | x == "", character(0),
                             strsplit(x, "/", fixed = TRUE))
  gset <- function(er) er@result %>%
    as_tibble() %>% transmute(path_num = normalize_pathway(ID),
                              gs = map(geneID, split_slash))

  genesets_all <- bind_rows(gset(ora_kegg_gene),
                            gset(ora_kegg_prot),
                            gset(ora_kegg_cpd)) %>%
    group_by(path_num) %>%
    summarise(geneSets = list(unique(unlist(gs))),
              Count    = lengths(geneSets),
              .groups  = "drop")

  sig2 <- sig %>% left_join(genesets_all, by = "path_num")

  # 5b) Rank pathways and keep top 40 ----------------------------------------
  sig2_ranked <- sig2 %>%
    mutate(
      nOmics = (!is.na(gene_padj)) + (!is.na(prot_padj)) + (!is.na(met_padj))
    ) %>%
    arrange(desc(nOmics), p_comb_stouffer_adj)

  top40_ids <- head(sig2_ranked$path_num, 40)

  # 5c) Build enrichResult object --------------------------------------------
  res_df <- sig2 %>% transmute(
    ID          = path_num,
    Description = description,
    pvalue      = p_comb_stouffer,
    p.adjust    = p_comb_stouffer_adj,
    qvalue      = p_comb_stouffer_adj,
    geneID      = map_chr(geneSets, paste, collapse = "/"),
    Count,
    GeneRatio   = paste0(Count, "/", Count),  # placeholders
    BgRatio     = GeneRatio
  ) %>% as.data.frame()

  er <<- new("enrichResult",
             result        = res_df,
             geneSets      = setNames(sig2$geneSets, sig2$path_num),
             pvalueCutoff  = 1, pAdjustMethod = "none",
             qvalueCutoff  = 1, organism = "UNKNOWN",
             keytype       = "UNKNOWN", readable = FALSE)

  # 5d) Add Jaccard similarity + relabel IDs ---------------------------------
  gs_list <- setNames(
    map(er@result$geneID,
        \(s) strsplit(s, "/", fixed = TRUE)[[1]] |> unique()),
    er@result$ID)

  ids <- names(gs_list); n <- length(ids)
  jac <- matrix(0, n, n, dimnames = list(ids, ids))
  for (i in seq_len(n)) {
    gi <- gs_list[[i]]
    for (j in i:n) {
      gj        <- gs_list[[j]]
      jac[i, j] <- jac[j, i] <-
        ifelse(length(unique(c(gi, gj))) == 0, 0,
               length(intersect(gi, gj)) / length(unique(c(gi, gj))))
    }
  }

  lab_vec <- with(er@result, setNames(Description, ID))
  rownames(jac)  <- colnames(jac)  <- lab_vec[rownames(jac)]
  names(gs_list) <- lab_vec[names(gs_list)]
  er@result$ID   <- lab_vec[er@result$ID]

  attr(er, "termsim") <- jac
  er@geneSets        <- gs_list

  # 5e) Draw treeplot ---------------------------------------------------------
  show_vec <- lab_vec[top40_ids] |> unname()

  cluster_cols <- c(
    "#2ca02c",  # cluster 1
    "#33caff",  # cluster 2
    "#9467bd",  # cluster 3
    "#d62727",  # cluster 4
    "#ffdf33"   # cluster 5
  )

  p <<- treeplot(
    er,
    showCategory   = show_vec,
    cluster.params = list(
      method = "average",
      n      = length(cluster_cols),
      color  = cluster_cols
    ),
    label_format   = 60,
    color          = "p.adjust"
  )

  ggsave(out_pdf, p, width = 16.5, height = 11.7, units = "in")

  ## 6) Heat‑map of “majority omics’’ per cluster -----------------------------
  if (verbose) message("➜ Step 6: building the majority‑omics heat‑map …")

  min_prop <- 0.75         # identifier must appear in ≥75 % of terms

  # 6a) TERM → raw identifiers
  term_to_ids <- setNames(
    map(er@result$geneID, \(s) unique(strsplit(s, "/", fixed = TRUE)[[1]])),
    er@result$ID
  )

  # 6b) Clusters present in the treeplot
  cluster_ids <- p[["data"]] |>
    filter(isTip, !is.na(group), grepl("^cluster_", group)) |>
    pull(group) |> unique()

  representative_ids_by_cluster <- list()
  for (cl in cluster_ids) {
    terms <- p[["data"]] |> filter(isTip, group == cl, !is.na(label)) |>
             pull(label) |> unique()
    if (!length(terms)) next
    ids_by_term <- term_to_ids[terms]
    freq <- table(unlist(ids_by_term))
    representative_ids_by_cluster[[cl]] <-
      sort(names(freq)[freq / length(ids_by_term) >= min_prop])
  }

  # 6c) Convert numeric gene IDs → miRNA; leave others unchanged
  geneid_to_symbol <- setNames(eg$SYMBOL, as.character(eg$ENTREZID))
  symbol_to_mirnas <- split(mirna_symbol$mature_mirna_id,
                            mirna_symbol$target_symbol)

  ids2miRNAs_keep_others <- function(ids) {
    ids           <- as.character(ids)
    id_is_numeric <- grepl("^\\d+$", ids)
    gene_ids      <- ids[id_is_numeric]
    other_ids     <- ids[!id_is_numeric]
    syms <- geneid_to_symbol[gene_ids] |> na.omit()
    miRs <- unique(unlist(symbol_to_mirnas[syms]))
    unique(c(other_ids, miRs))
  }

  representative_mixed_by_cluster <-
    lapply(representative_ids_by_cluster, ids2miRNAs_keep_others)

  # 6d) Long format dataframe for ggplot -------------------------------------
  df <- purrr::imap_dfr(representative_mixed_by_cluster,
                        ~tibble::tibble(id = .x, cluster = .y))

  # Rename KEGG compounds & UniProt IDs for readability -----------------------
  df$id <- as.character(df$id)
  id_is_kegg    <- grepl("^C\\d{5}$", df$id)              # KEGG compounds
  id_is_uniprot <- grepl("^[OPQ][0-9][A-Z0-9]{3}[0-9]$", df$id)  # UniProt 6‑char

  kegg_ids <- unique(df$id[id_is_kegg])
  prot_ids <- unique(df$id[id_is_uniprot])

  kegg_raw   <- vapply(kegg_ids, get_kegg_name,  character(1), USE.NAMES = TRUE)
  kegg_names <- sub(";.*$", "", kegg_raw) |> trimws()
  prot_names <- vapply(prot_ids, get_uniprot_name, character(1), USE.NAMES = TRUE)

  name_map <- c(kegg_names, prot_names)

  df <- df %>%
    mutate(id = if_else(id %in% names(name_map), name_map[id], id)) %>%
    mutate(
      id      = factor(id,      levels = rev(sort(unique(id)))),
      cluster = factor(cluster, levels = names(representative_mixed_by_cluster))
    )

  # 6e) Heat‑map palette & plotting ------------------------------------------
  cluster_cols_named <- setNames(cluster_cols, names(representative_mixed_by_cluster))

  heat_title <- paste0("Majority omics by cluster – ",
                       disease, " – ",
                       paste(modules, collapse = ", "))

  p_heat <- ggplot(df, aes(x = cluster, y = id)) +
    geom_tile(aes(fill = cluster), colour = "white", width = .9, height = .9) +
    scale_fill_manual(values = cluster_cols_named, guide = "none") +
    labs(title = heat_title, x = "Cluster", y = NULL) +
    theme_minimal(base_size = 12) +
    theme(
      axis.text.x = element_text(angle = 45, hjust = 1),
      panel.grid  = element_blank()
    )

  heatmap_pdf <- paste0(
    "majority_omics_by_cluster_",
    disease, "_",
    paste(modules, collapse = "_"),
    ".pdf"
  )
  ggsave(heatmap_pdf, p_heat, width = 10, height = 14, units = "in")
  if (verbose) message("✓ Heat‑map saved as \"", heatmap_pdf, "\".")

  ## 7) Return a list of useful objects --------------------------------------
  invisible(list(
    filtered_multiomics       = multi_omics_mod,
    ora_kegg_gene             = ora_kegg_gene,
    ora_kegg_prot             = ora_kegg_prot,
    ora_kegg_cpd              = ora_kegg_cpd,
    merged_kegg               = merged_kegg,
    sig_paths                 = sig2,
    enrichResult              = er,
    treeplot                  = p,
    representative_by_cluster = representative_mixed_by_cluster,
    majority_heatmap          = p_heat
  ))
}


```



```{r}
res <- multi_module_kegg_pipeline(
         modules = c("yellow","brown"),
         disease = "ibd",
         p_cutoff = 0.05,
         out_pdf  = "treeplot_multimod_ibd.pdf"
       )

```


